### ISSUE TITLE
Can I suspend and resume an windows process using this?

### ISSUE THREAD

**Issue by Cyberavater** [oshi/oshi#2482]
suspending and resuming the process in unix is very straightforward (SIGSTOP and SIGCONT) but in Windows, we need to call native API for that.

I was looking for something similar to [psutil](https://psutil.readthedocs.io/en/latest/#psutil.Process.suspend) in JVM, and oshi was the closest however I'm not sure if oshi can actually do this.

**Comment by dbwiddis** [comment]
> I was looking for something similar to [psutil](https://psutil.readthedocs.io/en/latest/#psutil.Process.suspend) in JVM, and oshi was the closest however I'm not sure if oshi can actually do this.

There is definitely a lot of functionality overlap with psutil (I've taken ideas from them and contributed some back) but we've had a fundamental principle here that OSHI is an "Information" library that collects information but does not provide methods for altering the OS or process/thread behavior.  

So the short answer is, "No, OSHI can't do this."  

But a longer answer is, "OSHI is JNA based (and uses the latest JNA as a dependency), and you can combine existing OSHI code with JNA methods which implement the native behavior."   Many Windows functions are already mapped in `jna-platform` and when they aren't, it's pretty easy to map your own.

I peeked at the [psutil implementation for this](https://github.com/giampaolo/psutil/blob/8eb2930539e8030bae94255040a6254889c0fb05/psutil/arch/windows/proc.c#L553-L556) which uses undocumented Windows functions `NTSuspendProcess` and `NTResumeProcess`.  There's another philosophical difference; we try to avoid non-API functions whenever possible.

There appears to be a workaround with Windows API Debug functions such as [DebugActiveProcess](https://learn.microsoft.com/en-us/windows/win32/api/debugapi/nf-debugapi-debugactiveprocess?redirectedfrom=MSDN).  These are not (yet) mapped in JNA (there's a feature request to do so, just waiting on someone in the community to contribute them...  https://github.com/java-native-access/jna/issues/1256 ... maybe that's you!). Happy to help you create those mappings yourself and it'd be great if you could also contribute them to JNA so others can benefit.

**Comment by Cyberavater** [comment]
> OSHI is an "Information" library that collects information but does not provide methods for altering the OS or process/thread behavior.

Does it at least support process [status?](https://psutil.readthedocs.io/en/latest/#psutil.WindowsService.status)

> I peeked at the [psutil implementation for this](https://github.com/giampaolo/psutil/blob/8eb2930539e8030bae94255040a6254889c0fb05/psutil/arch/windows/proc.c#L553-L556) which uses undocumented Windows functions NTSuspendProcess and NTResumeProcess. There's another philosophical difference; we try to avoid non-API functions whenever possible.

I've done the same using JNA (for my project), but haven't tested it yet. I was looking forward to making a toggle function (alternate suspend resume) but that would require knowing if it's currently suspended. So far what I found is [NtQuerySystemInformation()](https://stackoverflow.com/a/2881318/13519865) and a basic [implementation](https://stackoverflow.com/a/22949726/13519865)
But, I'm too dumb to implement this in JNA. (Again for unix it's pretty straightforward, I can just use `ps` stat and use `SIGCONT` if `T`.

**Comment by dbwiddis** [comment]
> Does it at least support process [status?](https://psutil.readthedocs.io/en/latest/#psutil.WindowsService.status)

Yep: https://javadoc.io/doc/com.github.oshi/oshi-core/latest/oshi/software/os/OSProcess.html#getState()


Yep: https://javadoc.io/doc/com.github.oshi/oshi-core/latest/oshi/software/os/OSProcess.html#getState() which should return SUSPENDED processes on Windows https://javadoc.io/static/com.github.oshi/oshi-core/6.4.6/oshi/software/os/OSProcess.State.html#SUSPENDED

However, note that this actually requires checking the thread state of each process: https://javadoc.io/doc/com.github.oshi/oshi-core/latest/oshi/software/os/OSProcess.html#getThreadDetails() and you can get the status of each thread https://javadoc.io/doc/com.github.oshi/oshi-core/latest/oshi/software/os/OSThread.html#getState() 

Note that there are some performance implications for this if you want it for all processes vs. single processes, for which there is a relevant config switch:
```
# Whether to update the OSProcess state on Windows to SUSPENDED if all its
# threads are suspended. This requires querying thread states and can impact
# performance (Process list queries can take much longer) but if desired is
# better done once than for each process. Users may still determine this value
# themselves by querying thread details, but this method is extremely slow if
# done for every process. Enabling this will provide a more efficient (but
# still slow) update if this information is of value. Defaults to false.
oshi.os.windows.procstate.suspended=false
```

> I've done the same using JNA (for my project), but haven't tested it yet. I was looking forward to making a toggle function (alternate suspend resume) but that would require knowing if it's currently suspended. So far what I found is [NtQuerySystemInformation()](https://stackoverflow.com/a/2881318/13519865) and a basic [implementation](https://stackoverflow.com/a/22949726/13519865) 

See discussion on https://github.com/oshi/oshi/issues/1581 for why we didn't use `NtQuerySystemInformation` but our existing implementation gives the same results.

> But, I'm too dumb to implement this in JNA. 

JNA is far easier than you might expect.  The linked JNA issue above actually contains exactly the interface you need to write!  But I'll improve on it a bit here:
```java
import com.sun.jna.win32.Kernel32;

public interface MyKernel32 extends Kernel32 {
    MyKernel32 INSTANCE = Native.load("Kernel32", MyKernel32.class, W32APIOptions.DEFAULT_OPTIONS);

    boolean DebugActiveProcess(int pid);
    boolean DebugSetProcessKillOnExit(boolean killOnExit);
    boolean DebugActiveProcessStop(int processId);
}
```
Then in your main code you just call `MyKernel32.INSTANCE.DebugActiveProcess(pid)` to suspend and `MyKernel32.INSTANCE.DebugActiveProcessStop(pid)` to resume.  (Add checking return values and error handling as needed.)

Let me know if you get this working and if you do I'd love to help you contribute this to JNA.

**Comment by Cyberavater** [comment]
Hello, thanks again for the long and helpful feedback I appreciate that.

> Yep: [javadoc.io/doc/com.github.oshi/oshi-core/latest/oshi/software/os/OSProcess.html#getState()](https://javadoc.io/doc/com.github.oshi/oshi-core/latest/oshi/software/os/OSProcess.html#getState()) which should return SUSPENDED processes on Windows [javadoc.io/static/com.github.oshi/oshi-core/6.4.6/oshi/software/os/OSProcess.State.html#SUSPENDED](https://javadoc.io/static/com.github.oshi/oshi-core/6.4.6/oshi/software/os/OSProcess.State.html#SUSPENDED)

Here's what I did:
```kotlin
fun main() {
    val systemInfo = SystemInfo()

    val os = systemInfo.operatingSystem

    val pid = 1228

    val process = os.getProcess(pid)

    if (process != null) {
        val state = process.state
        println("pid:$pid state:$state")
    } else {
        println("pid:$pid does not exist")
    }
}
```
Output: `pid:1228 state:RUNNING`
which is not expected since the Task Manager shows suspended.

python psutil also confirms this:
```python
import psutil

# pid = 28016
pid = 1228

process = psutil.Process(pid)
process.suspend()
print(process.status())
```
Output: stopped

> Note your program will need to have the appropriate debug permissions.
Does it mean any process that my program started will have this permission? BTW, I personally don't have anything against using `NTSuspendProcess` and `NTResumeProcess` even if it's undocumented. 
I've read about the most well-known [methods](https://stackoverflow.com/a/11010508/13519865)

- `SuspendThread` and `ResumeThread` can cause deadlock if not handled correctly.
- `DebugActiveProcessStop` would need its instance running to keep the process suspended (or it'll be terminated). Maybe setting `DebugSetProcessKillOnExit` to false is the way to go?
- `Command Line Utility` is not a portable option.

Thus I've concluded that `NTSuspendProcess` and `NTResumeProcess` are what would be best fit for me (also the JNA were not that complicated to write). But, I'm open to giving `DebugActiveProcessStop` a try at least (But again, this might be less efficient as 
I don't need the overhead of attaching a debugger if I simply want to pause it, correct me if I'm wrong).

**Comment by dbwiddis** [comment]
> Output: `pid:1228 state:RUNNING`
> which is not expected since the Task Manager shows suspended.

Yes, this is what I stated above; you have to enable the config setting to get Windows to report SUSPENDED at the process level. It requires iterating over every thread to check its state and is a massive performance hit if you do it for every process.

Use the Thread Details and iterate over the threads and check their state, or set the `oshi.os.windows.procstate.suspended` setting to `true`.

> DebugActiveProcessStop would need its instance running to keep the process suspended (or it'll be terminated). Maybe setting DebugSetProcessKillOnExit to false is the way to go?

I think so!  I've never tried this myself and don't have access to a Windows machine to test it.

>  I've concluded that NTSuspendProcess and NTResumeProcess are what would be best fit for me (also the JNA were not that complicated to write)

Yep, see how easy JNA is? :)

**Comment by dbwiddis** [comment]
> > Note your program will need to have the appropriate debug permissions.
> 
> Does it mean any process that my program started will have this permission?

This is a privilege your program needs to have.  If you're using OSHI as a dependency and running with elevated permissions it'll do that assignment for you the first time you call `systemInfo.getOperatingSystem()`. Otherwise you would need to do [this code](https://github.com/oshi/oshi/blob/11dee927ba3e11c74546631f28d20e72c3fe5a5e/oshi-core/src/main/java/oshi/software/os/windows/WindowsOperatingSystem.java#L440-L471) yourself.

**Comment by dbwiddis** [comment]
> But again, this might be less efficient as
> I don't need the overhead of attaching a debugger if I simply want to pause it, correct me if I'm wrong).

Well, your own program that calls this would be "the debugger" but I'm sure there is some overhead of configuring the process to send events. It just has the advantage that it's documented. :)

**Comment by Cyberavater** [comment]
> Use the Thread Details and iterate over the threads and check their state, or set the oshi.os.windows.procstate.suspended setting to true.

Do you mean `GlobalConfig.set("oshi.os.windows.procstate.suspended", true)' before checking the state? But, this gives me same result.

For suspending, understood, I appreciate your help.

**Comment by Cyberavater** [comment]
Okay now it works correctly,

```kotlin
fun main() {
    GlobalConfig.set("oshi.os.windows.procstate.suspended", true)
    
    val systemInfo = SystemInfo()

    val os = systemInfo.operatingSystem
    val pid = 1228

    val process = os.getProcess(pid)


    if (process != null) {
        val state = process.state
        println("pid:$pid state:$state")
    } else {
        println("pid:$pid does not exist")
    }
}
```

I had to set it even before `SystemInfo()` call

**Comment by dbwiddis** [comment]
> Do you mean `GlobalConfig.set("oshi.os.windows.procstate.suspended", true)` before checking the state? But, this gives me the same result.

Yes that's what I mean, but if you tweak config settings like this you need to do so at startup. They are read the first time you access a process listing and not re-loaded.  If you wait until just before checking state, it might be stuck at the default.

If you are only checking a single process, you should just iterate the `getThreadDetails()` states.  That's all the process-level one is doing anyway.
```java
int pid = this.getProcessID();
// If any thread is NOT suspended, set running
for (ThreadPerformanceData.PerfCounterBlock tcb : threadMap.values()) {
    if (tcb.getOwningProcessID() == pid) {
        if (tcb.getThreadWaitReason() == 5) {
            this.state = SUSPENDED;
        } else {
            this.state = RUNNING;
            break;
        }
    }
}
```

**Comment by Cyberavater** [comment]
I think all my problems are solved now. Thank you so much for your time and patience.
I'll close this issue now.

**Comment by dbwiddis** [comment]
> I had to set it even before `SystemInfo()` call

That works and is the recommended option.

Technically speaking you had up until the first `getProcess...()` call.

**Comment by Cyberavater** [comment]
Thanks again. Also,

> Of note, your question linked to psutil implementation of service status

My bad, I actually meant to link [process.status()](https://psutil.readthedocs.io/en/latest/#psutil.Process.status).
Sorry for the confusion.

**Comment by Cyberavater** [comment]
Okay, I've just done that

```kotlin
fun windowsProcessState(pid: Int) =
    ThreadPerformanceData.buildThreadMapFromRegistry(listOf(pid)).values.all {
        it.threadWaitReason == 5
    }.let {
        if (it) {
            OSProcess.State.SUSPENDED
        } else {
            OSProcess.State.RUNNING
        }
    }
```

it took arround `825.586700ms` while the previous method took around `1.061458001s`.

BTW, since I'll need oshi only for this specific fuction, maybe it'd be best if I just copy this portion, can you point me to where I should be looking in the codebase? (I find it somewhat hard tracing the nessary fucntions, 1. list all threads of a process, 2. get wait reason of a thread).

Thanks again for your help.

**Comment by dbwiddis** [comment]
> it took arround 825.586700ms while the previous method took around 1.061458001s.

Wow, that's still pretty long.  

> since I'll need oshi only for this specific fuction, maybe it'd be best if I just copy this portion, can you point me to where I should be looking in the codebase?

It's not quite trivial. There are three potential sources for this information.  The one you just used, is using the registry HKEY_PERFORMANCE_DATA.  The problem is that it fetches the registry blob for ALL the threads on the system, even if you just need one process.  So if you have a lot of processes it's ugly. It might actually be faster to query the PID threads using performance counters/WMI.  

You'll find the files you used in `oshi.driver.windows.registry`, the one that reads the Thread data is `ThreadPerformanceData` and it calls `HkeyPerformanceDataUtil`.

That same class has a method `buildThreadMapFromPerfCounters()` that you might try instead and see if it's a lot faster, as it'll only add the counters for the process you want.  You'll find the supporting files for that in the `oshi.driver.windows.perfmon` package, `ThreadInformation` is the one used and it calls `PerfCounterWildcardQuery` which is in the `oshi.util.platform.windows` package.

There may be a few util methods and interfaces you need outside of those, but most of the classes were committed upstream to JNA (jna-platform).  But OSHI isn't that big of a dependency, compared to all the JNA overhead you're bringing in...

**Comment by Cyberavater** [comment]
Thanks again. 
BTW, `buildThreadMapFromPerfCounters()` took around 3 seconds for a single pid.

**Comment by dbwiddis** [comment]
> BTW, `buildThreadMapFromPerfCounters()` took around 3 seconds for a single pid.

Yikes!  Something's gotta be wrong here.... 

... and I see we don't use the Perfmon API at all, we just query WMI for the info, again for ALL threads and not limited to just the specified process.

We can definitely make that better.  I'm going to create a new issue, I can write the code but don't have a Windows machine to test on.  Hopefully you'll help me test/benchmark it.

**Comment by Cyberavater** [comment]
> Hopefully you'll help me test/benchmark it.
I can do that. My CPU is i5 13th 13600K
![Screenshot 2023-10-21 010650](https://github.com/oshi/oshi/assets/32187927/7a53ec73-44bb-4a24-ade4-de65fc48821e)

Although I want to learn many things from you, you should know that I have very limited experience in programming, so you might need to spend more time than expected while working with me (I think you've guessed it already, but clarifying anyway).

**Comment by dbwiddis** [comment]
Excellent!    

No worries on the coding, and "contributing to open source" doesn't have to be coding -- just testing code on your hardware that I can't is extremely valuable.

I'll try an initial attempt at #2483 this weekend.  Are you able to checkout code from git and execute it locally?  Far easier to iterate on solutions that way than trying to get snapshots published (although we could go that route if necessary)

**Comment by Cyberavater** [comment]
>No worries on the coding, and "contributing to open source" doesn't have to be coding -- just testing code on your hardware that I can't is extremely valuable.

Thank you, I'll do my best.

>I'll try an initial attempt at https://github.com/oshi/oshi/issues/2483 this weekend. Are you able to checkout code from git and execute it locally? Far easier to iterate on solutions that way than trying to get snapshots published (although we could go that route if necessary)

Yes, I think I can checkout the code from git and test it. I also have experience using snapshot (which was published in Jitpack).

**Comment by dbwiddis** [comment]
Looks pretty good so far...
```
[INFO] Trying the old way:
[INFO]   Total time: 805ms
[INFO] Trying the new way:
[INFO]   Total time: 47ms
```

**Comment by dbwiddis** [comment]
Hey @Cyberavater, I've updated the code in https://github.com/oshi/oshi/pull/2487 so the "getThreadDetails" code uses the much shorter code path.  I've pushed it to snapshot (which should stay the same unless another PR gets merged before you look at it) or you can clone my branch where it derives from https://github.com/dbwiddis/oshi/tree/perfmon-thread or you can use the "gh" command line tool to locally checkout that PR.

TLDR:
 - from an API point of view, `getThreadDetails()` for a process will be much faster
 - from a "hack the backend class" point of view, use `buildProcessMapFromPerfCounters(Collection<Integer> pids, String procName)` using the process pid (as you did before) and the process name to filter the perfmon queries.

**Comment by Cyberavater** [comment]
I don't know if I did anything wrong, but I downloaded the jar and directly linked it. But I'm getting this error:

```
Exception in thread "main" java.lang.NoClassDefFoundError: org/slf4j/LoggerFactory
	at oshi.util.FileUtil.<clinit>(FileUtil.java:42)
	at oshi.util.GlobalConfig.<clinit>(GlobalConfig.java:23)
	at com.example.simpletests.MyClassKt.main(MyClass.kt:150)
	at com.example.simpletests.MyClassKt.main(MyClass.kt)
Caused by: java.lang.ClassNotFoundException: org.slf4j.LoggerFactory
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
	... 4 more

```

**Comment by dbwiddis** [comment]
If you're running from jar you also need to add the dependencies to the classpath.  In this particular case, SLF4J (slf4j-api) and you'll probably get a similar error for jna and jna-platform dependencies.

**Comment by dbwiddis** [comment]
You might find it easier to just point your program to the SNAPSHOT version:
https://stackoverflow.com/questions/7715321/how-to-download-snapshot-version-from-maven-snapshot-repository

**Comment by Cyberavater** [comment]
Here is my code:

```kotlin

fun windowsProcessState(pid: Int) =
    ThreadPerformanceData.buildThreadMapFromRegistry(listOf(pid)).also {
        if (it.isEmpty()) throw IllegalArgumentException("No thread data found for process $pid")
    }.values.all {
        it.threadWaitReason == 5
    }.let {
        if (it) {
            OSProcess.State.SUSPENDED
        } else {
            OSProcess.State.RUNNING
        }
    }

fun windowsProcessState2(pid: Int) =
    ThreadPerformanceData.buildThreadMapFromPerfCounters(listOf(pid)).also {
        if (it.isEmpty()) throw IllegalArgumentException("No thread data found for process $pid")
    }.values.all {
        it.threadWaitReason == 5
    }.let {
        if (it) {
            OSProcess.State.SUSPENDED
        } else {
            OSProcess.State.RUNNING
        }
    }

fun main() {
    val pid = 2164

    val t2 = measureTime {
        val state = windowsProcessState(pid)
        println("pid:$pid state:$state")
    }

    println(t2)
    
    val t1 = measureTime {
        GlobalConfig.set("oshi.os.windows.procstate.suspended", true)

        val systemInfo = SystemInfo()

        val os = systemInfo.operatingSystem


        val process = os.getProcess(pid)


        if (process != null) {
            val state = process.state
            println("pid:$pid state:$state")
        } else {
            println("pid:$pid does not exist")
        }
    }

    println(t1)



    val t3 = measureTime {
        val state = windowsProcessState2(pid)
        println("pid:$pid state:$state")
    }

    println(t3)

}
```

Here is my ouput:
```
pid:2164 state:SUSPENDED
487.610800ms
pid:2164 state:SUSPENDED
157.658900ms
pid:2164 state:SUSPENDED
3.939537900s
```

One thing that I noticed is that if I test using `getProcess` first buildThreadMapFromRegistry takes only 35ms

**Comment by dbwiddis** [comment]
OK, glad you got it working.  Some notes:
 - the registry method is what takes a long time and is always used for getting a process.  So setting `GlobalConfig.set("oshi.os.windows.procstate.suspended", true)` is making all your results slow. That really should only be set if you're regularly quering all processes for status (like a task manager equivalent)
 - the fastest approach is to leave that config setting at its (false) default, and simply query the process, iterating over its `getThreadDetails()` and checking if all are suspended.

**Comment by dbwiddis** [comment]
> if I test using `getProcess` first buildThreadMapFromRegistry takes only 35ms

OSHI uses memoizers, so with that config setting on, the first "getProcess" saves the output, and you're getting a cached value.

**Comment by Cyberavater** [comment]
Okay, here is my updated result:
```kotlin
fun main() {
    val pid = 2164

    val t2 = measureTime {
        val state = windowsProcessState(pid)
        println("pid:$pid state:$state")
    }

    println(t2)

//    val t1 = measureTime {
//        GlobalConfig.set("oshi.os.windows.procstate.suspended", true)
//
//        val systemInfo = SystemInfo()
//
//        val os = systemInfo.operatingSystem
//
//
//        val process = os.getProcess(pid)
//
//
//        if (process != null) {
//            val state = process.state
//            println("pid:$pid state:$state")
//        } else {
//            println("pid:$pid does not exist")
//        }
//    }
//
//    println(t1)



    val t3 = measureTime {
        val state = windowsProcessState2(pid)
        println("pid:$pid state:$state")
    }

    println(t3)

}
```

Output:
```
pid:2164 state:SUSPENDED
517.560200ms
pid:2164 state:SUSPENDED
4.084789900s
```

**Comment by Cyberavater** [comment]
Should I update these fuctions too?
```kotlin
fun windowsProcessState(pid: Int) =
    ThreadPerformanceData.buildThreadMapFromRegistry(listOf(pid)).also {
        if (it.isEmpty()) throw IllegalArgumentException("No thread data found for process $pid")
    }.values.all {
        it.threadWaitReason == 5
    }.let {
        if (it) {
            OSProcess.State.SUSPENDED
        } else {
            OSProcess.State.RUNNING
        }
    }

fun windowsProcessState2(pid: Int) =
    ThreadPerformanceData.buildThreadMapFromPerfCounters(listOf(pid)).also {
        if (it.isEmpty()) throw IllegalArgumentException("No thread data found for process $pid")
    }.values.all {
        it.threadWaitReason == 5
    }.let {
        if (it) {
            OSProcess.State.SUSPENDED
        } else {
            OSProcess.State.RUNNING
        }
    }

```

**Comment by dbwiddis** [comment]
The PerfCounters version needs to be modified to take arguments:
```java
buildThreadMapFromPerfCounters(Collection<Integer> pids, String procName, int threadNum) 
```
The one you're using still collects all the processes.

The only way to filter the list is including the process name (`osProcess.getName()`) as a second argument. Use -1 for the thread num to get all threads.

**Comment by dbwiddis** [comment]
Or just use the API, it's simpler:
 - don't override the config
 - get the process
 - iterate getThreadDetails 

It handles all the rest of the above code for you

**Comment by Cyberavater** [comment]
```
fun main() {
    val pid = 2164

//    val t2 = measureTime {
//        val state = windowsProcessState(pid)
//        println("pid:$pid state:$state")
//    }
//
//    println(t2)

    val t1 = measureTime {
//        GlobalConfig.set("oshi.os.windows.procstate.suspended", true)

        val systemInfo = SystemInfo()

        val os = systemInfo.operatingSystem


        val process = os.getProcess(pid)

        if (process != null) {
            val state = process.threadDetails.also {
                if (it.isEmpty()) throw IllegalArgumentException("No thread data found for process $pid")
            }.all {
                it.state == OSProcess.State.SUSPENDED
            }.let {
                if (it) {
                    OSProcess.State.SUSPENDED
                } else {
                    OSProcess.State.RUNNING
                }
            }
            println("pid:$pid state:${state}")
            

//                .forEach {
//                println("pid:$pid state:${it.state}")
//            }
        } else {
            println("pid:$pid does not exist")
        }

//        if (process != null) {
//            val state = process.state
//            println("pid:$pid state:$state")
//        } else {
//            println("pid:$pid does not exist")
//        }
    }

    println(t1)


//    val t3 = measureTime {
//        val state = windowsProcessState2(pid)
//        println("pid:$pid state:$state")
//    }
//
//    println(t3)

}
```
Output:
```
pid:2164 state:SUSPENDED
600.090100ms
```

I've run 3 tests and they're all consistent (all are around 600ms).

**Comment by Cyberavater** [comment]
```kotlin
fun main() {
    val pid = 2164

//    val t2 = measureTime {
//        val state = windowsProcessState(pid)
//        println("pid:$pid state:$state")
//    }
//
//    println(t2)

    val t1 = measureTime {
//        GlobalConfig.set("oshi.os.windows.procstate.suspended", true)

        val systemInfo = SystemInfo()

        val os = systemInfo.operatingSystem


        val process = os.getProcess(pid)

        if (process != null) {
//            val state =
            process.threadDetails.also {
                if (it.isEmpty()) throw IllegalArgumentException("No thread data found for process $pid")
            }.forEach {
                println("pid:$pid state:${it.state}")
            }

        } else {
            println("pid:$pid does not exist")
        }

//        if (process != null) {
//            val state = process.state
//            println("pid:$pid state:$state")
//        } else {
//            println("pid:$pid does not exist")
//        }
    }

    println(t1)


//    val t3 = measureTime {
//        val state = windowsProcessState2(pid)
//        println("pid:$pid state:$state")
//    }
//
//    println(t3)

}
```

Output:
```
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
621.409100ms
```

Here I printed all thread states.

**Comment by dbwiddis** [comment]
Looks about right.  You are capturing the time of the whole initialization as well, the process threads are probably a much shorter time.

**Comment by Cyberavater** [comment]
```kotlin
fun main() {
    val pid = 2164

//    val t2 = measureTime {
//        val state = windowsProcessState(pid)
//        println("pid:$pid state:$state")
//    }
//
//    println(t2)

    val systemInfo = SystemInfo()

    val os = systemInfo.operatingSystem


    val t1 = measureTime {
//        GlobalConfig.set("oshi.os.windows.procstate.suspended", true)

        val process = os.getProcess(pid)

        if (process != null) {
//            val state =
            process.threadDetails.forEach {
                println("pid:$pid state:${it.state}")
            }

        } else {
            println("pid:$pid does not exist")
        }

//        if (process != null) {
//            val state = process.state
//            println("pid:$pid state:$state")
//        } else {
//            println("pid:$pid does not exist")
//        }
    }

    println(t1)


//    val t3 = measureTime {
//        val state = windowsProcessState2(pid)
//        println("pid:$pid state:$state")
//    }
//
//    println(t3)

}
```
Output:
```
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
pid:2164 state:SUSPENDED
400.958100ms
```

This was  also consistent among 3 tests (all in ~400 ms).

**Comment by dbwiddis** [comment]
GetProcess still has to fetch all the processes so that's part of your time. 

Assuming you already had the process and just wanted to check the threads it's less. 

Or you can try the direct registry query with the process name (which you need to get the process to have!)

**Comment by Cyberavater** [comment]
Ah, so this
```kotlin
val process = os.getProcess(pid)
```
still takes time.

so, I should approach with this right?

>The PerfCounters version needs to be modified to take arguments:
buildThreadMapFromPerfCounters(Collection<Integer> pids, String procName, int threadNum) 
The one you're using still collects all the processes.
The only way to filter the list is including the process name (osProcess.getName()) as a second argument. Use -1 for the thread num to get all threads.

**Comment by dbwiddis** [comment]
> so, I should approach with this right?

Yes if you know the process name.

**Comment by Cyberavater** [comment]
```kotlin
fun main() {
    val pid = 2164

//    val t2 = measureTime {
//        val state = windowsProcessState(pid)
//        println("pid:$pid state:$state")
//    }
//
//    println(t2)

    val systemInfo = SystemInfo()

    val os = systemInfo.operatingSystem

    val process = os.getProcess(pid)


    val t1 = measureTime {
//        GlobalConfig.set("oshi.os.windows.procstate.suspended", true)


        if (process != null) {
//            val state =
            val isSuspended =
                process.threadDetails.filterNot { it.state == OSProcess.State.SUSPENDED }
                    .isEmpty()
            if (isSuspended) {
                println("The process is suspended")
            } else {
                println("The process is Running")
            }

//                .forEach {
//                println("pid:$pid state:${it.state}")
//            }

        } else {
            println("pid:$pid does not exist")
        }

//        if (process != null) {
//            val state = process.state
//            println("pid:$pid state:$state")
//        } else {
//            println("pid:$pid does not exist")
//        }
    }

    println(t1)


//    val t3 = measureTime {
//        val state = windowsProcessState2(pid)
//        println("pid:$pid state:$state")
//    }
//
//    println(t3)

}
```
Output:
```
The process is suspended
191.286ms
```

Yes if you know the process name.
Ah, yes. I still need  `os.getProcess(pid)` for the process name, since I only know pid.

**Comment by dbwiddis** [comment]
That's probably about the best you can get unless you really dig into the performance counter code to limit the query to only the info you want.

**Comment by Cyberavater** [comment]
Okay, thanks. I hope my work is done here, bye.

**Comment by dbwiddis** [comment]
Yep! And thanks for your help.
